src/metabase/prometheus.clj

```diff
@@ -0,0 +1,127 @@
+                    (GarbageCollectorExports.))
+   (collector/named {:namespace "metabase_application"
+                     :name      "jvm_standard"}
+                    (StandardExports.))
+   (collector/named {:namespace "metabase_application"
+                     :name      "jvm_memory_pools"}
+                    (MemoryPoolsExports.))
+   (collector/named {:namespace "metabase_application"
+                     :name      "jvm_threads"}
+                    (ThreadExports.))])
+
+(defn- jetty-collectors
+  []
+  [(JettyStatisticsCollector. (.getHandler (server/instance)))])
+
+(defn- setup-metrics!
+  "Instrument the application. Conditionally done when some setting is set. If [[prometheus-server-port]] is not set it
+  will throw."
+  [registry-name]
+  (log/info (trs "Starting prometheus metrics collector"))
+  (let [registry (prometheus/collector-registry registry-name)]
+    (apply prometheus/register registry
+           (concat (jvm-collectors)
+                   (jetty-collectors)))))
+
+(defn- start-web-server!
+  "Start the prometheus web-server. If [[prometheus-server-port]] is not set it will throw."
+  [port registry]
+  (log/info (trs "Starting prometheus metrics web-server on port {0}" port))
+  (when-not port
+    (throw (ex-info (trs "Attempting to set up prometheus metrics web-server with no web-server port provided")
+                    {})))
+  (ring-jetty/run-jetty (-> (constantly {:status 200})
+                            (collector.ring/wrap-metrics registry {:path "/metrics"}))
+                        {:join?       false
+                         :port        port
+                         :max-threads 8}))
+
+(p.types/defprotocol+ PrometheusActions
+  (stop-web-server [this]))
+
+(p/defrecord+ PrometheusSystem [registry web-server]
+  PrometheusActions
+  (stop-web-server [_this]
+    (when-let [^Server web-server web-server]
+      (.stop web-server))))
+
+(defn- make-prometheus-system
+  "Takes a port (zero for a random port in test) and a registry name and returns a [[PrometheusSystem]] with a registry
+  serving metrics from that port."
+  [port registry-name]
+  (let [registry (setup-metrics! registry-name)
+        web-server (start-web-server! port registry)]
+    (->PrometheusSystem registry web-server)))
+
+(defn setup!
+  "Start the prometheus metric collector and web-server."
+  []
+  (let [port (prometheus-server-port)]
+    (when-not port
+      (throw (ex-info (trs "Attempting to set up prometheus metrics with no web-server port provided")
+                      {})))
+    (when-not system
+      (locking #'system
+        (when-not system
+          (let [sys (make-prometheus-system port "metabase-registry")]
+            (alter-var-root #'system (constantly sys))))))))
+
+(defn shutdown!
+  "Stop the prometheus metrics web-server if it is running."
+  []
+  (when system
+    (try (stop-web-server system)
+         (log/info (trs "Prometheus web-server shut down"))
```

Comment:
Should the' system be reset to `nil` after shutting down?

I can see this might make development that involves starting/stopping Prometheus more transparent. especially when the work involves configuring prometheus.